src/optimal_image/search.rs

```diff
@@ -1,35 +1,127 @@
-use dataclients::ImageData;
+use dataclients::*;
+use image::*;
+use std::collections::HashMap;
+use std::error::Error;
 use std::path::Path;
+extern crate dssim;
+use EncodingConfig;
 use ImageConfig;
 use ImageFormat;
 use ImageScore;
 
+#[derive(Debug)]
 pub struct SearchOptions {
-    threshold: f64,
-    formats: Vec<ImageFormat>,
+    pub threshold: f64,
+    pub quality_range: (u8, u8),
+    pub formats: Vec<ImageFormat>,
 }
 
+pub type SearchResult = HashMap<ImageFormat, ImageConfig>;
+
 pub struct Search {
+    path: String,
```

Comment:
What's the diff between `path` and `image_data`?

Reply:
have to figure out if it is the best interface here, but path would be the path to the original image, used to read it and generate the image_data. I kept it in a field as a possible output of the search may be the image itself, but could also be the image path plus the optimal config options to be generated elsewhere.